persistent/cPersistence.c

```diff
@@ -973,7 +973,9 @@ Per__p_getattr(cPersistentObject *self, PyObject *name)
 }
 
 /*
-  TODO:  we should probably not allow assignment of __class__ and __dict__.
+ * TODO:  we should probably not allow assignment of __class__ and __dict__.
+ * But we do, and have for a long time. Code in the wild is known to rely on
+ * assigning __class__.
```

Comment:
Sure. One example is these [two](https://github.com/NextThought/nti.site/blob/bf6980eb8ee63b8c12979d498fec1b1dc7c7f3fe/src/nti/site/site.py#L392-L398) related [parts]( https://github.com/NextThought/nti.site/blob/bf6980eb8ee63b8c12979d498fec1b1dc7c7f3fe/src/nti/site/site.py#L441-L448) of `nti.site`.

Basically, the `Components` object of `zope.interface` implements `__bases__` by having its tree of objects copy references to the two adapter registries that make up a `Components`. This means that if we need to change the implementation of one of those adapter registries, we can't do so by just assigning to, e.g., `self.adapters`: there are likely many other references to the old object still around somewhere.  Thus if we want to make a change, the object has to be mutated *in place*.  For behaviour changes in methods, that means assigning to `__class__`.

This problem is magnified by persistence, especially because, while the adapter registries from `zope.component.persistentregistery` are `Persistent`, the `PersistentComponents` actually is *not*, making migration painful.

This is ultimately the motivation for https://github.com/zopefoundation/zope.interface/pull/228